presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java

```diff
@@ -140,16 +145,43 @@ private Patterns() {}
         return typeOf(WindowNode.class);
     }
 
-    public static <C> Property<PlanNode, C, PlanNode> source()
-    {
-        return optionalProperty("source", node -> node.getSources().size() == 1 ?
-                Optional.of(node.getSources().get(0)) :
-                empty());
-    }
-
-    public static <C> Property<PlanNode, C, List<PlanNode>> sources()
-    {
-        return property("sources", PlanNode::getSources);
+    public static Explore<PlanNode, Rule.Context, PlanNode> source()
+    {
+        return explore(
+                "source",
+                (node, context) -> {
+                    if (node.getSources().size() != 1) {
+                        return Stream.of();
+                    }
+                    return context.getLookup().resolveGroup(getOnlyElement(node.getSources()));
+                });
+    }
+
+    public static Explore<PlanNode, Rule.Context, List<PlanNode>> sources()
+    {
+        return explore(
+                "sources",
+                (PlanNode node, Rule.Context context) -> {
+                    List<Stream<PlanNode>> sourceStreams = node.getSources().stream()
+                            .map(source -> context.getLookup().resolveGroup(source))
+                            .collect(toImmutableList());
+
+                    return Streams.stream(new AbstractIterator<List<PlanNode>>() {
+                        @Override
+                        protected List<PlanNode> computeNext()
+                        {
+                            List<PlanNode> sources = sourceStreams.stream()
+                                    .map(source -> source.findFirst()
+                                            .orElse(null)).collect(toImmutableList());
+                            sourceStreams.stream().forEach(sourceStream -> sourceStream.skip(1));
+
+                            if (sources.contains(null)) {
+                                return endOfData();
+                            }
+                            return sources;
```

Comment:
You're basically making a matrix of possible source combinations  and transforming it (matrix-wise).

1. can you simply enumerate the possibilities and make this transformation simpler?
2. you arbitrarily combine first member in each group as one combination, second member of each group as another combination, etc... Is it correct, without making a cartesian product here? (which could obviously explode)
3. .. which boils down to conclusion: should you do this at all without actually changing iterative's logic? This starts to outgrow iterative optimizer's assumption